### DISCUSSION TITLE
text features in graph neural network with large number of data nodes

### DISCUSSION THREAD

**Discussion by icedpanda** [pyg-team/pytorch_geometric#5762] [Q&A] [answered]
Hi, I would like to use text features as node features for a GNN. 

I need to compute **all nodes embeddings per batch** in order to get all candidate embeddings for a downstream task. 

Similar to #2765 but I need to make the language model trainable. (which makes it challenging since a `LM` will take large memory)

What would be the reasonable approach to this?

This is my pseudocode

```python
# in a forward call
def forward():
    avg_text_emb = []
    # can't load all text to bert
    for text in all_text:
        # assume we have batch of 8
        # bert is unfreezed
        text_embedding = bert(text).pooler_output 
        # average embeddings
        avg_text_emb.append(text_embedding.mean(dim=1))
   # all candidate text embeddings
   # I have 20K nodes and 200K edges
   all_avg_text_emb = torch.stack(avg_text_emb)
  
   # all candidate embeddings from the gnn for a downstream task
   all_node_features = gnn(avg_text_emb, edge_idx)
 
   # downstream task
   score = some_task(all_node_features, node_features)
   return score
```

Any suggestion would be helpful. Thanks!.

**Comment by rusty1s** [comment, marked as answer]
I think your code looks reasonable but I don't think it will be possible to train a large language model end-to-end in combination with a GNN for around 20k nodes. Alternatives include
1. using a more light-weight language embedding model, e.g., bag-of-words or averaging Word2Vec embeddings
2. Pre-processing the intermediate and final outputs of BERT and using them as input features

**Reply by icedpanda** [reply]
Hi @rusty1s,

I ended up using K-candidates instead of using all node features. However, how can I pass the edge_index to a RGCN in this case? (edge_index mapping all candidate IDs)

```python
# assume these are all edge idx
edge_idx = [[0,1,2,3,4,5], [2,3,4,0,1]]
edge_type = [0,1,2,0,1,2]

#assume these embeddings are from the BERT output but use nn.Embedding for example
# 6 nodes, 
avg_text_emb = nn.Embedding(6, 5)

# now only use the sub-set of this edge_idx
k_edge_idx = [[1,2,4], [3,4,0]]
k_edge_type = [1,2,1]
k_text_emb = avg_text_emb(k_edge_idx[0])

# I don't think this is gonna to work as the index here is still mapping to the original index order.  
# for example node 4 should be pointing to k_text_emb[3] not  k_text_emb[4]
k_node_features = rgcn(x=k_text_emb, k_edge_idx, k_edge_type)

```
I hope this makes sense. Thansk

**Reply by rusty1s** [reply]
Yes, you would need to have a vector that maps the local indices in `k_edge_index` to its global node indices. This is something you automatically get by running `utils.subgraph()` with `relabel_nodes=True`.

**Reply by icedpanda** [reply]
is there any function that can create neighbors based on the given `node_idx`? I check `NeighborSampler` but seems to be the data loader. I want to sample k-Neighbors during the fly based on the `node_idx` from my own data loader.

```python
def forward():
    #  batch_size, n nodex_idx
    node_idx = batch["node_idx"]
    # sample k neighbor, assume 10 neighbors for the first layer and 20 neighbors for the second layer
    # self.edge_idx has all edges
    # k_edge_idx is relabeled idx that needs to match with text_embed.
    # k_global_edge_idx is for retrieving the text
    k_edge_idx, edge_type, k_global_edge_idx = neighbor_sampler(self.edge_idx, self.edge_type, sizes=[10, 20],  node_idx=node_idx)
    # get unique node index to retrieve text
    k_nodes_idx,= unique(k_global_idx)
    k_text = self.all_text[k_nodes_idx]

    # can't load all text to bert
    for text in k_text:
        # assume we have a batch of 8
        # bert is unfreezed
        text_embedding = bert(text).pooler_output 
        # average embeddings
        avg_text_emb.append(text_embedding.mean(dim=1))

   avg_text_emb = torch.stack(avg_text_emb)
  
   k_node_features = r_gcn(avg_text_emb,  k_edge_idx, edge_type)

```

Thanks

**Reply by icedpanda** [reply]
the sampler from `torch_geometric.sampler` looks like the one I need, but I can't import it and its not in the `__init__.py` file

**Reply by rusty1s** [reply]
It got recently added and is only available on master. See [here](https://github.com/pyg-team/pytorch_geometric#nightly-and-master) on how to install.

**Reply by icedpanda** [reply]
hi @rusty1s, sorry to bother you again. 

How can I return neighbors using `sampler.NeighborSampler` from a list of nodes like `utils.subgraph()`? I managed to install the `nightly version`.

I want to return the neighbors of a list of nodes based on the given `node_index`. The `sampler.NeighborSampler.sample_from_nodes()` expects only 3 tensors as input, but I want to pass a list of nodes.

Hope you can help me with this. thanks in advance!

**Reply by rusty1s** [reply]
You can pass in a list of nodes as a tensor, e.g., `sample_from_nodes(torch.tensor([0, 1, 2, 3]))`.

**Reply by icedpanda** [reply]
```python
import torch
from torch_geometric.sampler import NeighborSampler
from torch_geometric.data import Data

edge_index = torch.tensor([
    [0, 1, 1, 2, 2, 3],
    [1, 0, 2, 1, 3, 2],
])

# only 1 neighbour
sampler = NeighborSampler(data, num_neighbors=[1])
out = sampler.sample_from_nodes(torch.tensor([1,2]))
# should expecting 1 neighbour from node_idx 1 and node_index 2
```
But got 
```
ValueError: not enough values to unpack (expected 3, got 2)
```

https://pytorch-geometric.readthedocs.io/en/latest/modules/sampler.html

**Reply by rusty1s** [reply]
Ah, you are right. The interface slightly changed. Try
```python
import torch

from torch_geometric.data import Data
from torch_geometric.sampler import NeighborSampler

edge_index = torch.tensor([
    [0, 1, 1, 2, 2, 3],
    [1, 0, 2, 1, 3, 2],
])

data = Data(edge_index=edge_index, num_nodes=4)

# only 1 neighbour
sampler = NeighborSampler(data, num_neighbors=[1])
example_idx = torch.arange(4)
node_idx = torch.tensor([1, 2, 3, 4])
out = sampler.sample_from_nodes((example_idx, node_idx, None))
```

**Reply by icedpanda** [reply]
what is `example_idx` here? 

I try to `sample_from_noes()` 5 times but always get the same neighbors. shouldn't it return different neighbors each time in a uniform way? (e.g. 1->0, 1->2, 3->2)

```python
import torch

from torch_geometric.data import Data
from torch_geometric.sampler import NeighborSampler

edge_index = torch.tensor([
    [0, 1, 1, 2, 2, 3],
    [1, 0, 2, 1, 3, 2],
])

data = Data(edge_index=edge_index, num_nodes=4)

# only 1 neighbour
sampler = NeighborSampler(data, num_neighbors=[1])
example_idx = torch.arange(4)
node_idx = torch.tensor([1, 3])

for i in range(5):
    out = sampler.sample_from_nodes((example_idx, node_idx, None))
    print(out)

# SamplerOutput(node=tensor([1, 3, 0, 2]), row=tensor([2, 3]), col=tensor([0, 1]), edge=tensor([1, 5]), batch=None, metadata=tensor([0, 1, 2, 3]))
# SamplerOutput(node=tensor([1, 3, 0, 2]), row=tensor([2, 3]), col=tensor([0, 1]), edge=tensor([1, 5]), batch=None, metadata=tensor([0, 1, 2, 3]))
# SamplerOutput(node=tensor([1, 3, 0, 2]), row=tensor([2, 3]), col=tensor([0, 1]), edge=tensor([1, 5]), batch=None, metadata=tensor([0, 1, 2, 3]))
# SamplerOutput(node=tensor([1, 3, 0, 2]), row=tensor([2, 3]), col=tensor([0, 1]), edge=tensor([1, 5]), batch=None, metadata=tensor([0, 1, 2, 3]))
# SamplerOutput(node=tensor([1, 3, 0, 2]), row=tensor([2, 3]), col=tensor([0, 1]), edge=tensor([1, 5]), batch=None, metadata=tensor([0, 1, 2, 3]))
``` 


btw, could you please point me out the document for the `SamplerOutput`? I can't find it in the document.

**Reply by rusty1s** [reply]
It's defined [here](https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/sampler/base.py#L9-L54).
